realtek: new dgs-1210 device tree and add support for dgs-1210-10mp

flygarn12 opened the pull request:
Commit topics: (2022-08-04)

1. A new updated modular dgs-1210 family Device Tree that can more easily be expandable and configured to all devices in the family. This device tree will also be able to handle different SoC series in combination with common gpio and poe hardware everything more easily controlled from the dts file. Also implementing the new Device Tree for the dgs-1210 devices in OpenWrt.

2. Adding support for rtl8380 based D-Link DGS-1210-10MP.

flygarn12 commented:
I don't get what the test formalias actually want because I have made the pull request the same way as everyone else!?
And how do I restart the test when I have changed the headline and put < > around the mail address, but there was there before also but they didn't show up unless there was space added?
I have put my name in the signed off, just as everyone else so why isn't it approved?

flygarn12 commented:
`Author name (flygarn12) need to be your real name 'firstname lastname'`
What author name, and where do I change that?

flygarn12 commented:
`Commit subject line MUST start with '<area>: ' (Add support for dgs-1210-10mp)`
What? There isn't a single open commit that is named like this: '<Realtek>: ' (Add support for dgs-1210-10mp)` so what am I supposed to write as a headline?

flygarn12 commented:
`Signed-off-by is missing or doesn't match author (should be 'Signed-off-by: flygarn12 <87232422+flygarn12@users.noreply.github.com>')`
Why doesn't it see my signed off line that has been there all the time?

flygarn12 commented:
Have rewritten the commit with the actual commit message and changed the PR message to just add support of the device.

flygarn12 commented:
In the commit message: fixed some typos, adjusted strange tabs in the installation instruction and made some clarifications in the UART pinout.

paulfertser commented:
Hi @flygarn12 ,
Is it possible the device you have there has same GPIO arrangement as in https://www.mail-archive.com/openwrt-devel@lists.openwrt.org/msg60129.html ? I do not see you specifying any SFP-related settings, can you please check if my numbers make SFP hotplug and "ethtool -m" work on your board please?

flygarn12 commented:
Hi @paulfertser ,
It is right I haven't officially touched the I2C hardware (but I have a trigger finger on it!). I asked on the forum if anyone had tried to get it working since no one device has this function in the dts files so and the short answer I got was "no". 
Therefore I now decided to commit what I actually had working. 
But I personally believe your port numbers should work since everything else seem to be the same pin numbers on these 8+2 port devices. But I doesn't have any test hardware at this point to probe the SFP connections inside the cages. 
So my plan was to commit the device as it works and later continue making the I2C/SFP working. I also wanted to make the poe_max LED official so it should be able to be included in the realtek-poe project as the error output.

flygarn12 commented:
Do anyone know how to remove commits because when I updated my fork GitHub actually copied the OpenWrt updates to this PR and when I tried to revert it it now made 3 commits and 15 files changes !?

svanheule commented:
Instead of merging the OpenWrt master branch, you should rebase onto it. That will give conflicts when you try to push to your github mirror, but you just have to force push then.

flygarn12 commented:
Ok thanks @svanheule 
There are so many functions and traps to fall in to when using GitHub.
But now the unnecessary commits are gone and we are back on track again :-)

musashino205 commented:
Please use the actual target name for the prefix in the commit title, ```realtek: ```. You can also use ```realtek: rtl838x: ```.

svanheule commented:
> note GitHub add code function is really bad and it doesn't see line feeds!?

See Github's [Mardown docs](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code) on how to provide blocks of code. (You need to use triple back-ticks.)

flygarn12 commented:
Changed target name to Realtek and some minor changes in commit message hardware data.

svanheule commented on target/linux/realtek/dts-5.10/rtl8382_d-link_dgs-1210-10mp.dts:
```diff
@@ -0,0 +1,183 @@
+			partition@1c40000 {
+				label = "jffs2";
+				reg = <0x01c40000 0x3c0000>;
+			};
```
This partition table is identical to the one in `rtl8382_d-link_dgs-1210.dtsi`, so you can include that file and drop some things here. If this device has an RTL8380 SoC, the dtsi should be renamed first. (`rtl83xx_...` since the [DGS-1210-52](https://github.com/openwrt/openwrt/pull/10227) has the same common parts.)

flygarn12 replied:
I am building a completely redesigned family Device Tree.

svanheule commented on target/linux/realtek/image/rtl838x.mk:
```diff
@@ -27,6 +27,13 @@ define Device/d-link_dgs-1210
+define Device/d-link_dgs-1210-10mp
+  $(Device/d-link_dgs-1210)
+  DEVICE_MODEL := DGS-1210-10MP
+  DEVICE_PACKAGES += lua-rs232
```
This used to be included to use the PoE lua script, but a different package is currently being created and submitted for inclusion in OpenWrt. Please drop this package.

flygarn12 replied:
So this isn't needed for the realtek-poe package?

svanheule reviewed (changes requested):
The target name in the commit title is not capitalised: "realtek: ...".

Did you verify with the boot log that this chip is an RTL8382? Devices with less than 16 ports usually have an RTL8380.

Since you've added the "lua-rs232" package, I assume this device has BCM59xxx PSE chips, but it would be good to have this in the commit message too. Especially since this is a currently-unsupported feature, this can also help others in the future to select the right packages for this device.

For the MTBF, people can just consult the device datasheet. In my opinion, it's not very useful info to have in the commit message.

flygarn12 commented:
Yes it has the same PoE controllers as the -10 where the lua-rs232 actually came from.

How much hardware info is really meaningful to have in the commit message? Pretty much everything is device datasheet info anyway. When I look around on old commits there seem to be no standard what to include if to include hardware info at all?
Not even install info seem to be a demand, som include it some don't. Sometime the commit message (like for dgs-1210-10p) is only "here it is. Signed off by someone".
It is pretty much only the dgs-1210-16 device that actually have a install instruction that goes for all.

It feels sometime better to have install and hardware (at least hardware) in the wiki since that is editable afterward.

But the ER4 for example had install info in the commit and then something changed and the install procedure stoped working and got unofficially replaced by a forum post.

svanheule commented:
The "Test Formalities" have failed because you have "`flygarn12 <87232422+flygarn12@users.noreply.github.com>`" as the commit author. Please update this (`git commit --amend --reset-author`).

flygarn12 commented:
I tried a making a new totally revised version of the commit message with hopefully more useful information instead of something that looks like a copy of the D-Link datasheet. What do you think of this before I amend it (or throw it away)?

```
General hardware info
---------------------

D-Link DGS-1210-10MP is a switch with 8 ethernet ports and 2 SFP ports, all ports Gbit capable.
It is based on a RTL8380 SoC @ 500MHz, DRAM 128MB and 32MB Flash.
All ethernet ports are 802.3af/at PoE capable.

DTS and rtl838x.mk file info
----------------------------
I originally started with the dgs-1210-10p device.

rtl838x.mk:
 - lua-rs232 is included for the communication to the PoE hardware (BCM59121).

rtl8382_d-link_dgs-1210-10mp.dts:
 - LED's poe_max, link_act and poe is tested manually with luci by turning on the led.
 - gpio-restart is tested by disable wdt0 and activate the gpio1 34.
 - mode key is tested with reset function, 
   the mode key is in dts defined with linux,code = <KEY_LIGHTS_TOGGLE> but that code doesn't do anything.
 - reset key is tested and works. Fast push reboot the device and a 6-7second push reset the device.

UART for the SoC
----------------
UART for the SoC (115200 8N1) is available close to the front panel next to the LED/KEY card connector
via unpopulated standard 0.1" pin header marked J4.
Pin1 is marked with arrow and square.

Pin 1: Vcc 3,3V
Pin 2: Tx
Pin 3: Rx
Pin 4: Gnd

Installation with TFTP from U-Boot (check if newer procedure can be found in the OpenWrt device wiki)
-----------------------------------------------------------------------------------------------------
Note: If you want a genuine switch function and you have a master router with dhcp server you can make this 
      installation procedure simpler with a uci-defaults script with these three settings, then you don't 
      need to do installation steps 9, 10 and 15:
      uci del network.lan.ipaddr
      uci del network.lan.netmask
      uci set network.lan.proto='dhcp'
      uci commit network
	  
Note: The file name openwrt-realtek-rtl838x-d-link_dgs-1210-10mp-initramfs-kernel.bin can be shortened to initramfs.bin, for example.
	   
1. Prepare a TFTP server with the OpenWrt *initramfs-kernel.bin
2. Connect this network port to one of switch ports.
   Note! u-boot always have address of 10.90.90.90 with subnetmask 255.0.0.0.
   A network with these settings must be setup to proceed.
   The switch must be set to have static lease for this address in the network.
3. Connect to the serial console and power on the switch.
4. Press the ESC key once you see “Hit Esc key to stop autoboot” in the console output 
   (very short timeframe so I recommend to push the ESC from boot time until it stops).
5. Press CTRL+C keys to get into the real U-Boot prompt.
6. Init the network with the command: rtk network on
7. Load the OpenWrt image with the command: tftpboot 0x8f000000 <ipofyournetworkport>:<openwrtimagename>
   <ipofyournetworkport> is your TFTP server IP, e.g. 10.90.90.100
   <openwrtimagename> is the name of the image provided by the TFTP server (*initramfs-kernel.bin)
   Example: u-boot># tftpboot 0x8f000000 10.90.90.100:openwrt-realtek-rtl838x-d-link_dgs-1210-10mp-initramfs-kernel.bin
8. Boot the OpenWrt image with the command: bootm
9. Once booted, modify /etc/config/network:
   - (if needed!) remove the port, where your PC is connected to, from the wan_vlan network (option ports)
   - (if needed!) add that port to the lan_vlan network (option ports)
   - uci del network.lan.ipaddr
   - uci del network.lan.netmask
   - uci set network.lan.proto='dhcp'
   - uci commit network
10. Reload the network configuration with the command: /etc/init.d/network reload
11. Open LuCI.
12. Upload the the OpenWrt *squashfs-sysupgrade.bin to the switch.
13. Wait for it to reboot.
15. Repeat steps 9 and 10 via the serial access to get access from the network side, again.

Original OEM hardware/firmware data at first installation
---------------------------------------------------------
It has been installed, developed, and tested on a device with these OEM hardware and firmware versions.

- U-Boot: 2011.12.(2.1.5.67086)-Candidate1 (Jun 22 2020 - 15:03:58)
- Boot Version: 1.01.001
- Firmware Version: 6.20.007
- Hardware Version: F1

Things to be done when support are developed
--------------------------------------------
 - PoE is tested and works with the unofficial package realtek-poe. When this is implemented in OpenWrt
   the package should be included as device_packages in the rtl838x.mk.
 - LED link_act and poe are functional connected to the poe system. 
   Since the PoE package isn't official these LED can't do anything. But probably some software development is 
   also needed to shift the led array between act and poe indication and to shift the mode lights with mode key.
 - Led poe_max should probably be used as straight forward error output from the realtek-poe package error handling.
 - SFP is not currently hot pluggable. Development is under progress to get working I2C communication with SFP
   and have them hot pluggable. When any device in the dgs-1210 family gets this working, it should be possible to 
   implement the same solution in this device.

```

flygarn12 commented:
@svanheule do you think it could be a good idea if I make the general dtsi work in a separate PR since it will include small changes in the whole d-link family files since the include definitions must be updated with a new name for the dtsi?
And when that is finished come back to this device.

flygarn12 commented:
I am changing this PR to draft while I modifies and test the review comments and I am going to do a separate PR for the dtsi modifications that must be merged before this PR can be merged.

paulfertser commented:
On Fri, Jul 22, 2022 at 01:19:40PM -0700, Daniel Groth wrote:
> I am going to do a separate PR for the dtsi modifications that must be merged
> before this PR can be merged.

This idea is usually expressed by pushing a "Github Pull Request"
containing two or more commits in the right order.

flygarn12 commented:
> On Fri, Jul 22, 2022 at 01:19:40PM -0700, Daniel Groth wrote: I am going to do a separate PR for the dtsi modifications that must be merged before this PR can be merged.
> This idea is usually expressed by pushing a "Github Pull Request" containing two or more commits in the right order.
I actually just figured that out so that the is the path I am working on right now. It got so complicated to separate them since they must come together to actually work.

svanheule commented:
@flygarn12 I'm going to reduce the noise in my inbox a bit and unsubscribe here. Please ping me when your patches are ready for review. Thanks in any case for taking care of the clean-up :)

flygarn12 commented:
```
Conflicting files
target/linux/realtek/dts-5.10/rtl8382_d-link_dgs-1210.dtsi 
```
There is no conflict. This file has to be renamed and updated as the commit say if this device tree are supposed to work.

flygarn12 commented:
> ```
> Conflicting files
> target/linux/realtek/dts-5.10/rtl8382_d-link_dgs-1210.dtsi 
> ```
> 
> There is no conflict. This file has to be renamed and updated as the commit say if this device tree are supposed to work.

This has now been solved by updating the working branch.

RaylynnKnight commented on target/linux/realtek/dts-5.10/rtl8380_d-link_dgs-1210-10p.dts:
```diff
@@ -0,0 +1,79 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "rtl838x_d-link_dgs-1210_soc.dtsi"
+
```
Why are you missing the poe include here?  The DGS-1210-10P also has PoE with a 65W budget.

flygarn12 replied:
It isn't missing, it has never been defined in the original dts and have only made a 1:1 rebuild. Please see commit message part: 'Implementation:'.

flygarn12 replied:
This device SoC change has been removed from the commit.

RaylynnKnight commented on target/linux/realtek/dts-5.10/rtl8380_d-link_dgs-1210-10mp.dts:
```diff
@@ -0,0 +1,57 @@
+
+#include "rtl838x_d-link_dgs-1210_soc.dtsi"
+#include "rtl83xx_d-link_dgs-1210_poe.dtsi"
+
```
Shouldn't this also include the gpio related dtsi?

flygarn12 replied:
It does include the gpio dtsi. This is described in the commit message at note 6. Or just as easy in the top of the poe.dtsi, poe can't exist without the hardware in gpio.dtsi.

flygarn12 replied:
Also see the commit message part: 'How to implement in dts file settings:'

oliv3r replied:
I think @RaylynnKnight is trying to say 'include what you use'.

It's bad practice to rely on includes of other includes. What if the POE stuff doesn't include it anymore? Now your thing is broken as well.

flygarn12 replied:
Has been changed.

RaylynnKnight commented on target/linux/realtek/dts-5.10/rtl8380_d-link_dgs-1210-10p.dts:
```diff
@@ -0,0 +1,79 @@
+		indirect-access-bus-id = <0>;
+	};
+};
+
```
The section from line 9 to 26 should likely be replaced with #include "rtl83xx_d-link_dgs-1210_gpio.dtsi".

flygarn12 replied:
No, I don't do changes on unfinished devices or devices I can't test. I only make 1:1 changes. So all the devices that has been included in the OpenWrt source earlier will be exactly the same after my commits.
Please see commit message part: 'Implementation:'.

svanheule replied:
I agree that the initial patches should be no-ops, resulting in identical output. @flygarn12, I think @RaylynnKnight owns one of these -10p devices, so a Tested-by could be provided if you decided to write an extra follow-up patch. But then you would sort of be obliged to accommodate for everyone's requests for device updates, and that could delay the series. So I suggest we wait with these corrections until we get the details of the restructuring ironed out.

flygarn12 replied:
@svanheule @RaylynnKnight This patch is only a collection of what we know and what is official. I can't correct every non complete dts that has been already merged in this patch, that should have been handled before when those devices was merged at the beginning.

svanheule replied:
Of course you can't know all those details, and I don't expect you to correct any previous mistakes. You can keep the scope of this PR to what it is now. If people care enough about other devices, those fixes will get in at some point.

RaylynnKnight replied:
@flygarn12 OK, I understand your purpose for this patch-set now.  I was kind of hoping that @paulfertser effort could be combined with this to provide good support to more D-Link devices.  But once your patch is merged perhaps those with the other D-Link devices will have more incentive to fix missing functionality.

flygarn12 replied:
This commit has been removed from the PR.

RaylynnKnight commented on target/linux/realtek/dts-5.10/rtl8382_d-link_dgs-1210-16.dts:
```diff
@@ -1,6 +1,6 @@
 
-#include "rtl8382_d-link_dgs-1210.dtsi"
+#include "rtl838x_d-link_dgs-1210_soc.dtsi"
 
```
Any reason you did not include the gpio related dtsi here as well?

flygarn12 replied:
Yes. The answer is very much the same as your -10p question. This is how the hardware in the -16.dts file was defined and I don't do changes on the hardware in this project. I have only made a 1:1 rebuild so was it there before it will be there after.

RaylynnKnight reviewed (commented):
You are also missing the SFP related nodes which can be found in a previous RFC submitted by @paulfertser which can be found at https://patchwork.ozlabs.org/project/openwrt/patch/20211005194024.31909-3-fercerpav@gmail.com/

flygarn12 commented:
> You are also missing the SFP related nodes which can be found in a previous RFC submitted by @paulfertser which can be found at https://patchwork.ozlabs.org/project/openwrt/patch/20211005194024.31909-3-fercerpav@gmail.com/

It isn't missing. It doesn't work. Read the comments from the top.

oliv3r commented on target/linux/realtek/dts-5.10/rtl8380_d-link_dgs-1210-10mp.dts:
```diff
@@ -0,0 +1,57 @@
+#include "rtl83xx_d-link_dgs-1210_poe.dtsi"
+
+/ {
+	compatible = "d-link,dgs-1210-10mp", "realtek,rtl838x-soc";
```
```suggestion
	compatible = "d-link,dgs-1210-10mp", "realtek,rtl8382-soc", "realtek,rtl838x-soc";
```
allows you at least to have some quirck/overrides if needed be

flygarn12 replied:
This has been changed.

oliv3r commented on target/linux/realtek/dts-5.10/rtl8380_d-link_dgs-1210-10mp.dts:
```diff
@@ -0,0 +1,57 @@
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		INTERNAL_PHY(8)
```
ugh, I guess this target already uses these macro's, but I strongly dislike them personally. Just write it out, we write code once, but read it many times, being specific here, means it's not up for interpretation later.

flygarn12 replied:
@oliv3r I am not sure what you want me to do here?
```
#address-cells = <1>;
#size-cells = <0>;
```
Am I going to replace this with something and in that case what? Or am I supposed to remove it?

All devices in the family has exactly this code?

flygarn12 replied:
I have looked in every dts and dtsi file there is for all realtek devices and every device have exactly this code so there is no problem.

oliv3r commented on target/linux/realtek/dts-5.10/rtl8380_d-link_dgs-1210-10mp.dts:
```diff
@@ -0,0 +1,57 @@
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		SWITCH_PORT(8, 1, internal)
```
as above, it's not bad to be expressive here :)

flygarn12 replied:
This is also the standard code for every dts file in every realtek device so there is no problem.

oliv3r replied:
I understand, I know we do a lot of copy/pasting around here, so this is thus inherent from this. Doesn't make it any better or less ugly :) But this can easily be fixed once it goes upstream, and getting it merged now is better of course.

oliv3r commented on target/linux/realtek/dts-5.10/rtl8380_d-link_dgs-1210-10p.dts:
```diff
@@ -0,0 +1,79 @@
+		compatible = "gpio-keys-polled";
+		poll-interval = <20>;
+
+		/* is this pin 30 on the external RTL8231 (&gpio1)? */
```
left-over debug commit?

flygarn12 replied:
@oliv3r @musashino205 Do we talk about line 13? But this sounds like a problem in the current -10p device dts. My patch doesn't fix old problems in the current dts. It only makes a new device tree that works with every device based on the current facts. So the old input will be the new output, nothing more or nothing less.

flygarn12 replied:
This commit has been removed from the PR.

oliv3r commented on target/linux/realtek/dts-5.10/rtl838x_d-link_dgs-1210_soc.dtsi:
```diff
@@ -1,5 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0-or-later OR MIT
 
+// Includes d-link dgs-1210 SoC common functions.
```
```suggestion
/* Includes d-link dgs-1210 SoC common functions. */
```
Linus still prefers `/*` comments over `//` comments (bar the SDPX license identifier ...)

flygarn12 replied:
Comment has been removed.

oliv3r commented on target/linux/realtek/dts-5.10/rtl83xx_d-link_dgs-1210_gpio.dtsi:
```diff
@@ -0,0 +1,29 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+// Includes d-link dgs-1210 family common board functions connected to the rtl8231 GPIO expander, no PoE functions.
```
try to keep comments 80 columns aligned ...

flygarn12 replied:
Comment has been removed.

flygarn12 commented:
Ok, have now updated this PR with new commits.
I have made two big changes on top of the suggestions I got from the last review!
1. The SoC change from 8382 to 8380 for dgs-1210-10p has been removed. The owners of this device has to make some update work on the dts file where hopefully my new gpio and poe dtsi files will work well.
2. I found out yesterday that the realtek-poe package is now official so that is now included for the dgs-1210-10mp. My suggestion for a working poe config file is in the commit message, but this has to be written/updated manually as this point.

flygarn12 commented:
> The target name in the commit title is not capitalised: "realtek: ...".
> 
> Did you verify with the boot log that this chip is an RTL8382? Devices with less than 16 ports usually have an RTL8380.
> 
> Since you've added the "lua-rs232" package, I assume this device has BCM59xxx PSE chips, but it would be good to have this in the commit message too. Especially since this is a currently-unsupported feature, this can also help others in the future to select the right packages for this device.
> 
> For the MTBF, people can just consult the device datasheet. In my opinion, it's not very useful info to have in the commit message.

@svanheule this is locked as a change requested, but I think I have changed this but the lock still stands.

svanheule commented on target/linux/realtek/dts-5.10/rtl83xx_d-link_dgs-1210_poe.dtsi:
```diff
@@ -0,0 +1,30 @@
+	poe {
+		label = "green:poe";
+		gpios = <&gpio1 29 GPIO_ACTIVE_LOW>;
+	};
```
It's not entirely clear to me how these work. Am I correct to understand from your commit message that each port has one of these LEDs? Or are these global status LEDs?

If these GPIOs hook up to a shift register, you can add an extra GPIO node for that, to expose all the LEDs. See https://elixir.bootlin.com/linux/v5.10.135/source/Documentation/devicetree/bindings/gpio/gpio-74x164.txt

flygarn12 replied:
Answer is in comments below.

svanheule reviewed (commented):
Thanks for the hard work, this looks pretty good!

One minor nit on the DTS rework, commented inline. The `_poe.dtsi` file is also new in the rework commit, but you only use it on the 10MP. You only have to add that file in the second commit (building on your new file structure), so the changes can be more contained.

Please also take care to leave the file modes to 644 (RW for owner, R for others), since now you have them set to 755 (Windows defaults?). I can also fix this when merging.

Last, but not least, I would like to ask you to provide a real email address for Author and Signed-off-by. I know Github provides these addresses for privacy, but the commit message tags are there specifically to be able to contact you later on. A no-reply address defeats that goal, so I'm afraid we can't accept these. If you're afraid about getting spammed or really want keep your personal address private (which I fully understand), you could create a new mailbox or alias (depending on you current mail host).

flygarn12 commented:
@svanheule 
```
One minor nit on the DTS rework, commented inline. The _poe.dtsi file is also new in the rework commit, but you only use it on the 10MP. You only have to add that file in the second commit (building on your new file structure), so the changes can be more contained.

```
This hardware with poe vs link led mode with mode key I pretty much expect to be the same on every PoE device but we only have one single poe device now in OpenWrt and that is -10p and that dts isn't complete. But I would be surprised if that differ much from -10mp.
![image](https://user-images.githubusercontent.com/87232422/183278490-622eaf89-d58f-46b0-a97f-ea0e1faa8631.jpeg)

svanheule commented:
Aha, so those LEDs are an indication of the current mode of the port LEDs. Thanks for clarifying.

I still think the `_poe.dtsi` should be included with the first device that uses it, but the way you split things it also has its merits (adding a family-wide include to the family-wide DTS rework).

flygarn12 commented:
> Aha, so those LEDs are an indication of the current mode of the port LEDs. Thanks for clarifying.
> 
> 
> 
> I still think the `_poe.dtsi` should be included with the first device that uses it, but the way you split things it also has its merits (adding a family-wide include to the family-wide DTS rework).

Yes, I have been thinking about this question this morning. As you say, it is probably better to have the specific poe hardware specified in the -10mp dts until we have at least two devices (more ports devices than -10p and -10mp which are kind of brothers in the d-link family) with poe so we see the trend of port connections.
Because the point of _poe.dtsi do pretty much fall apart if most devices doesn't use the same ports. The _gpio.dtsi was easier to do because a lot of hardware had already mapped the gpio chip. The anomaly there is the dgs-1210-16, in the manual it has a reset key but it isn't specified in the dts!?

But these changes might take some days to do since I am a bit occupied daytime for these coming weeks.

svanheule commented:
> The anomaly there is the dgs-1210-16, in the manual it has a reset key but it isn't specified in the dts!?

I think the DGS-1210-16 was one of the first to be merged, so don't worry about it if things appear to be missing. Most likely just an oversight.

> But these changes might take some days to do since I am a bit occupied daytime for these coming weeks.

No worries, take your time. We'll get there eventually :)

flygarn12 committed:
realtek: d-link: dgs-1210 remake and implement a new device tree

I have collected the known information from the dts files we have at this
date and I looked at the D-Link manual for this family
series to get some view on what LED and KEY setup that we can expect.

I have then made a new Device Tree that should work for this whole family with
maybe little tweaks in the future when we learn more of the devices.
This device tree is based on modules where you first select which SoC group the
device belongs to. Then you include ether the gpio or poe dtsi file depending
on what hardware your device has, see examples below.

-------------------------------------------------------------------------------
The device tree looks like this:
	----------------			----------------
	| rtl838x.dtsi |			| rtl839x.dtsi |  // Note 1.
	----------------			----------------
		|					|
		|					|
-------------------------------------	-------------------------------------
|rtl838x_d-link_dgs-1210_common.dtsi|	|rtl839x_d-link_dgs-1210_common.dtsi|
-------------------------------------	-------------------------------------
		|					|	// Note 2.
		|_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _|
				|
				|	--------------
				|-------| device.dts |	// Note 3.
				|	--------------
				|
		------------------------------------
		| rtl83xx_d-link_dgs-1210_gpio.dtsi |	// Note 4.
		------------------------------------
				|
				|	--------------
				|-------| device.dts |	// Note 5.
				|	--------------
Current options			|
- - - - - - - - - - - - - - - - |- - - - - - - - - - - - - - - - - -
Future options			|
				|
		------------------------------------
		| rtl83xx_d-link_dgs-1210_poe.dtsi |	// Note 6.
		------------------------------------
				|
				|	--------------
				|-------| device.dts |	// Note 7.
					--------------

Note 1; Included in rtl838x_d-link_dgs-1210_common.dtsi and
	rtl839x_d-link_dgs-1210_common.dtsi.

Note 2; MCU and SoC level information and memory mapping. Choose which one to include
	in the device dts.
	Note that the rtl839x_d-link_dgs-1210_common.dtsi is future option.

Note 3; At this point dgs-1210-16 will come out here, will maybe move to note 5
	in the future if more features is added in the dts file.

Note 4; In this dtsi only common board hardware based on the rtl8231 is found.
	No PoE based hardware.
	In this dtsi there is no <#include> to above *_common.dtsi.

Note 5; Device dts with only rtl8231 based hardware without PoE will come out
	here.

Note 6 and 7 is a future option!
Note 6; In this dtsi only common hardware with PoE will be found.
	The poe dtsi must be included in the device dts.

Note 7; Device with PoE will come out here.

-------------------------------------------------------------------------------
How to set up in dts file:

The device dts have one of these three <#include> alternatives.

This alternative includes only common features:
<#include "rtl838x_d-link_dgs-1210_common.dtsi">

This alternative include common and the (non PoE) rtl8231 GPIO features:
<#include "rtl838x_d-link_dgs-1210_common.dtsi">
<#include "rtl83xx_d-link_dgs-1210_gpio.dtsi">

Future option:
This alternative include common, rtl8231 GPIO and PoE features.
<#include "rtl838x_d-link_dgs-1210_common.dtsi">
<#include "rtl83xx_d-link_dgs-1210_gpio.dtsi">
<#include "rtl83xx_d-link_dgs-1210_poe.dtsi">

-------------------------------------------------------------------------------
Implementation:

Finally, I also implementing this new family device tree on the
current supported devices:
dgs-1210-10p
dgs-1210-16
dgs-1210-20
dgs-1210-28

The implementation for the dgs-1210-10p is different. I have removed the
information from the rtl8382_d-link_dgs-1210-10p.dts that are in common with
the rtl838x_d-link_dgs-1210_common.dtsi.
Since the rest isn't officially probed in the device dts I do not want to
include the rtl83xx_d-link_dgs-1210_gpio.dtsi with dgs-1210-10p.dts.
Someone with the hardware device must test this hardware with
rtl83xx_d-link_dgs-1210_gpio.dtsi.

Since I don't have these devices to test on I have built the original firmware
for each one of these devices before this change and saved the dtb file and
then compared the original dtb file with the dtb file built with this new
device tree.

-------------------------------------------------------------------------------
What is not included at this point:

As for now the rtl839x_d-link_dgs-1210_common.dtsi doesn't exist, the first
device we have in his family that use this SoC will be dgs-1210-52.
But to fully implement it to the tree we probably need more devices based on
rtl839x or else we only split one single dts to two dtsi files plus one dts.

rtl83xx_d-link_dgs-1210_poe.dtsi since we only have one port probed PoE device
this dtsi have not been included this time. When we get more PoE devices we
can make this dtsi file.

I looked at the port and switch setup to see if it could be moved to the dtsi.
I decided not to touch this part now. The reason was that there isn't really
any meaningful way this could be shared between the devices.
The only thing in common over the family is the 8+2sfp ports on the
dgs-1210-10xx device, but they don't start at number one.
And then there is the hot plug SFP and I2C ports that aren’t implemented
on any device. So maybe when we see the whole ethernet map for the family
then maybe the ports can be moved to a *_common.dtsi but I don't think it is
the right moment for that now.

Signed-off-by: Daniel Groth <flygarn12@gmail.com>

flygarn12 committed:
realtek: d-link: add support for dgs-1210-10mp

General hardware info
-------------------------------------------------------------------------------

D-Link DGS-1210-10MP is a switch with 8 ethernet ports and 2 SFP ports, all
ports Gbit capable. It is based on a RTL8380 SoC @ 500MHz, DRAM 128MB and
32MB Flash. All ethernet ports are 802.3af/at PoE capable.

File info
-------------------------------------------------------------------------------
The dgs-1210-10mp is very similar to dgs-1210-10p so I used that as a start.

rtl838x.mk:
 - Removed lua-rs232 package since it was a leftover from the old rtl83xx-poe
   package.
 - Updated the SOC to 8380.
 - Installed the new realtek-poe package.

rtl8380_d-link_dgs-1210-10mp.dts:
 - Moved dgs-1210 family common parts and non poe related ports on rtl8231
   to the new device tree dtsi files.

UART for the SoC
-------------------------------------------------------------------------------
UART for the SoC (115200 8N1) is available close to the front panel next to the
LED/KEY card connector via unpopulated standard 0.1" pin header marked J4.
Pin1 is marked with arrow and square.

Pin 1: Vcc 3,3V
Pin 2: Tx
Pin 3: Rx
Pin 4: Gnd

Installation with TFTP from U-Boot
(for updated procedure check OpenWrt device wiki)
-------------------------------------------------------------------------------
1. Prepare a TFTP server with the OpenWrt *initramfs-kernel.bin
2. Connect this network port to one of switch ports.
   Note! u-boot always have address of 10.90.90.90 with subnetmask 255.0.0.0.
   A network with these settings must be setup to proceed.
   The switch must be set to have static lease for this address in the network.
3. Connect to the serial console and power on the switch.
4. Press the ESC key once you see “Hit Esc key to stop autoboot” in the console
   output.
5. Press CTRL+C keys to get into the real U-Boot prompt.
6. Init the network with the command: rtk network on
7. Load the OpenWrt image with the command:
   tftpboot 0x8f000000 <ipofyournetworkport>:<openwrtimagename>
   <ipofyournetworkport> is your TFTP server IP, e.g. 10.90.90.100
   <openwrtimagename> is the name of the image provided by the TFTP server
   (*initramfs-kernel.bin)
8. Boot the OpenWrt image with the command: bootm
9. Once booted, modify /etc/config/network:
   - (if needed!) remove the port, where your PC is connected to, from the
     wan_vlan network (option ports)
   - (if needed!) add that port to the lan_vlan network (option ports)
   - uci del network.lan.ipaddr
   - uci del network.lan.netmask
   - uci set network.lan.proto='dhcp'
   - uci commit network
10. Reload the network configuration with the command:
    /etc/init.d/network reload
11. Open LuCI.
12. Upload the the OpenWrt *squashfs-sysupgrade.bin to the switch.
13. Wait for it to reboot.
15. Repeat steps 9 and 10 via the serial access to get access from the
    network side, again.

About the realtek-poe package
-------------------------------------------------------------------------------
The realtek-poe package is installed but there isn't any automatic config
setting at this time. Therefore the poe config file has to be edited manually
for this device.

Since the device has a total PoE budget of 130W I suggest replacing the
contents of file /etc/config/poe with this:

config global
	option budget	'130'

config port
	option enable	'1'
	option id	'1'
	option name	'lan1'
	option poe_plus	'1'
	option priority	'1'

config port
	option enable	'1'
	option id	'2'
	option name	'lan2'
	option poe_plus	'1'
	option priority	'1'

config port
	option enable	'1'
	option id	'3'
	option name	'lan3'
	option poe_plus	'1'
	option priority	'1'

config port
	option enable	'1'
	option id	'4'
	option name	'lan4'
	option poe_plus	'1'
	option priority	'1'

config port
	option enable	'1'
	option id	'5'
	option name	'lan5'
	option poe_plus	'1'
	option priority	'1'

config port
	option enable	'1'
	option id	'6'
	option name	'lan6'
	option poe_plus	'1'
	option priority	'1'

config port
	option enable	'1'
	option id	'7'
	option name	'lan7'
	option poe_plus	'1'
	option priority	'1'

config port
	option enable	'1'
	option id	'8'
	option name	'lan8'
	option poe_plus	'1'
	option priority	'1'

Original OEM hardware/firmware data at first installation
-------------------------------------------------------------------------------
It has been installed, developed, and tested on a device with these OEM
hardware and firmware versions.

- U-Boot: 2011.12.(2.1.5.67086)-Candidate1 (Jun 22 2020 - 15:03:58)
- Boot Version: 1.01.001
- Firmware Version: 6.20.007
- Hardware Version: F1

Things to be done when support are developed
-------------------------------------------------------------------------------
 - realtek-poe has been included in OpenWrt but the automatic config handling
   has not been solved yet so in the future there will probably be some minor
   updates for this device to handle the poe config.
 - LED link_act and poe are per function supposed to be connected to the poe
   system.
   But some software development is also needed to make this led work and
   shift the led array between act and poe indication and to shift the mode
   lights with mode key.
 - Led poe_max should probably be used as straight forward error output from
   the realtek-poe package error handling. But no code has been written for
   this.
 - SFP is currently not hot pluggable. Development is under progress to get
   working I2C communication with SFP and have them hot pluggable.
   When any device in the dgs-1210 family gets this working, I expect it
   should be possible to implement the same solution in this device.

Signed-off-by: Daniel Groth <flygarn12@gmail.com>

flygarn12 commented:
Ok made some changes and have now updated the PR with this changes:
1. Change commit email.
2. Removed the _poe.dtsi and moved the poe information back to dgs-1210-10mp.dts.
3. Generally changed file permissions back to 644 from 755.
4. Changed name on the file rtl838x_d-link_dgs-1210_soc.dtsi to rtl838x_d-link_dgs-1210_common.dtsi.
5. Updated commit messages to reflect the new changes.

svanheule commented on target/linux/realtek/image/rtl838x.mk:
```diff
@@ -27,6 +27,14 @@ define Device/d-link_dgs-1210
+define Device/d-link_dgs-1210-10mp
+  $(Device/d-link_dgs-1210)
+  SOC := rtl8380
+  DEVICE_MODEL := DGS-1210-10MP
```
Please add `DEVICE_VARIANT := F`.

Only the recent family revisions use Realtek SoCs, so I think we should be specific about this one.

svanheule reviewed (commented):
First commit message:
```
> I have collected the known information from the dts files we have at this
> date and I looked at the D-Link manual for this family
> series to get some view on what LED and KEY setup that we can expect.
> 
> I have then made a new Device Tree that should work for this whole family with
> maybe little tweaks in the future when we learn more of the devices.
> This device tree is based on modules where you first select which SoC group the
> device belongs to. Then you include ether the gpio or poe dtsi file depending
```
"include either"
```
> on what hardware your device has, see examples below.
> 
> -------------------------------------------------------------------------------
> The device tree looks like this:
> 	----------------			----------------
> 	| rtl838x.dtsi |			| rtl839x.dtsi |  // Note 1.
> 	----------------			----------------
> 		|					|
> 		|					|
> -------------------------------------	-------------------------------------
> |rtl838x_d-link_dgs-1210_common.dtsi|	|rtl839x_d-link_dgs-1210_common.dtsi|
> -------------------------------------	-------------------------------------
> 		|					|	// Note 2.
> 		|_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _|
> 				|
> 				|	--------------
> 				|-------| device.dts |	// Note 3.
> 				|	--------------
> 				|
> 		------------------------------------
> 		| rtl83xx_d-link_dgs-1210_gpio.dtsi |	// Note 4.
> 		------------------------------------
> 				|
> 				|	--------------
> 				|-------| device.dts |	// Note 5.
> 				|	--------------
> Current options			|
> - - - - - - - - - - - - - - - - |- - - - - - - - - - - - - - - - - -
> Future options			|
> 				|
> 		------------------------------------
> 		| rtl83xx_d-link_dgs-1210_poe.dtsi |	// Note 6.
> 		------------------------------------
> 				|
> 				|	--------------
> 				|-------| device.dts |	// Note 7.
> 					--------------
> 
> Note 1; Included in rtl838x_d-link_dgs-1210_common.dtsi and
> 	rtl839x_d-link_dgs-1210_common.dtsi.
> 
> Note 2; MCU and SoC level information and memory mapping. Choose which one to include
```
Which MCU? At this point, you only have the SoC details, no extra microcontroller. Did you mean CPU (from rtl83{8x,9x}.dtsi)?
```
> 	in the device dts.
> 	Note that the rtl839x_d-link_dgs-1210_common.dtsi is future option.
```
"is a future option"
```
> 
> Note 3; At this point dgs-1210-16 will come out here, will maybe move to note 5
> 	in the future if more features is added in the dts file.
```
"which will maybe move"
"if more feature are added"
```
> 
> Note 4; In this dtsi only common board hardware based on the rtl8231 is found.
> 	No PoE based hardware.
> 	In this dtsi there is no <#include> to above *_common.dtsi.
> 
> Note 5; Device dts with only rtl8231 based hardware without PoE will come out
> 	here.
> 
> Note 6 and 7 is a future option!
> Note 6; In this dtsi only common hardware with PoE will be found.
> 	The poe dtsi must be included in the device dts.
> 
> Note 7; Device with PoE will come out here.
> 
> -------------------------------------------------------------------------------
> How to set up in dts file:
> 
> The device dts have one of these three <#include> alternatives.
```
"will have", "has", or "DTSs". (The latter would require the whole commit message to be updated for consistent capitalisation of abbreviations) 
```
> 
> This alternative includes only common features:
> <#include "rtl838x_d-link_dgs-1210_common.dtsi">
> 
> This alternative include common and the (non PoE) rtl8231 GPIO features:
```
"includes"
```
> <#include "rtl838x_d-link_dgs-1210_common.dtsi">
> <#include "rtl83xx_d-link_dgs-1210_gpio.dtsi">
> 
> Future option:
> This alternative include common, rtl8231 GPIO and PoE features.
```
"includes"
```
> <#include "rtl838x_d-link_dgs-1210_common.dtsi">
> <#include "rtl83xx_d-link_dgs-1210_gpio.dtsi">
> <#include "rtl83xx_d-link_dgs-1210_poe.dtsi">
> 
> -------------------------------------------------------------------------------
> Implementation:
> 
> Finally, I also implementing this new family device tree on the
```
"implemented"
```
> current supported devices:
> dgs-1210-10p
> dgs-1210-16
> dgs-1210-20
> dgs-1210-28
> 
> The implementation for the dgs-1210-10p is different. I have removed the
> information from the rtl8382_d-link_dgs-1210-10p.dts that are in common with
> the rtl838x_d-link_dgs-1210_common.dtsi.
```
"the information from rtl8382_d-link_dgs-1210-10p.dts that is shared with rtl838x_d-link_dgs-1210_common.dtsi"
or "that is already present in rtl838x_d-link_dgs-1210_common.dtsi"
```
> Since the rest isn't officially probed in the device dts I do not want to
> include the rtl83xx_d-link_dgs-1210_gpio.dtsi with dgs-1210-10p.dts.
> Someone with the hardware device must test this hardware with
> rtl83xx_d-link_dgs-1210_gpio.dtsi.
> 
> Since I don't have these devices to test on I have built the original firmware
> for each one of these devices before this change and saved the dtb file and
> then compared the original dtb file with the dtb file built with this new
> device tree.
> 
> -------------------------------------------------------------------------------
> What is not included at this point:
> 
> As for now the rtl839x_d-link_dgs-1210_common.dtsi doesn't exist, the first
> device we have in his family that use this SoC will be dgs-1210-52.
> But to fully implement it to the tree we probably need more devices based on
> rtl839x or else we only split one single dts to two dtsi files plus one dts.
> 
> rtl83xx_d-link_dgs-1210_poe.dtsi since we only have one port probed PoE device
> this dtsi have not been included this time. When we get more PoE devices we
> can make this dtsi file.
> 
> I looked at the port and switch setup to see if it could be moved to the dtsi.
> I decided not to touch this part now. The reason was that there isn't really
> any meaningful way this could be shared between the devices.
> The only thing in common over the family is the 8+2sfp ports on the
> dgs-1210-10xx device, but they don't start at number one.
> And then there is the hot plug SFP and I2C ports that aren’t implemented
> on any device. So maybe when we see the whole ethernet map for the family
> then maybe the ports can be moved to a *_common.dtsi but I don't think it is
> the right moment for that now.
```

Second commit message:
```
> Installation with TFTP from U-Boot
> (for updated procedure check OpenWrt device wiki)
> -------------------------------------------------------------------------------
> 1. Prepare a TFTP server with the OpenWrt *initramfs-kernel.bin
> 2. Connect this network port to one of switch ports.
>    Note! u-boot always have address of 10.90.90.90 with subnetmask 255.0.0.0.
>    A network with these settings must be setup to proceed.
>    The switch must be set to have static lease for this address in the network.
> 3. Connect to the serial console and power on the switch.
> 4. Press the ESC key once you see “Hit Esc key to stop autoboot” in the console
>    output.
> 5. Press CTRL+C keys to get into the real U-Boot prompt.
> 6. Init the network with the command: rtk network on
> 7. Load the OpenWrt image with the command:
>    tftpboot 0x8f000000 <ipofyournetworkport>:<openwrtimagename>
>    <ipofyournetworkport> is your TFTP server IP, e.g. 10.90.90.100
>    <openwrtimagename> is the name of the image provided by the TFTP server
>    (*initramfs-kernel.bin)
> 8. Boot the OpenWrt image with the command: bootm
> 9. Once booted, modify /etc/config/network:
>    - (if needed!) remove the port, where your PC is connected to, from the
>      wan_vlan network (option ports)
>    - (if needed!) add that port to the lan_vlan network (option ports)
>    - uci del network.lan.ipaddr
>    - uci del network.lan.netmask
>    - uci set network.lan.proto='dhcp'
>    - uci commit network
```
You could also specify the default address "192.168.1.1" and "change the network configuration as required". But this probably helps new users, so it can stay.
```
> 10. Reload the network configuration with the command:
>     /etc/init.d/network reload
> 11. Open LuCI.
> 12. Upload the the OpenWrt *squashfs-sysupgrade.bin to the switch.
> 13. Wait for it to reboot.
> 15. Repeat steps 9 and 10 via the serial access to get access from the
>     network side, again.
> 
> About the realtek-poe package
> -------------------------------------------------------------------------------
> The realtek-poe package is installed but there isn't any automatic config
> setting at this time. Therefore the poe config file has to be edited manually
> for this device.
> 
> Since the device has a total PoE budget of 130W I suggest replacing the
> contents of file /etc/config/poe with this:
> 
[file contents]
```
If the commit message contains all the required info for a user to create a correct config file, this verbatim copy isn't required. I think this belongs more on a device wiki page.

Neustradamus commented:
To follow this PR :)

svanheule commented:
I've merged these changes with updated commit messages from Daniel:
51ec9b386472 realtek: d-link: dgs-1210 remake of the device tree
8c04a5c45677 realtek: d-link: add support for dgs-1210-10mp
